### PR TITLE
Add Blueprint System for Task Composition and Reusability

### PR DESCRIPTION
**Pull request by theandrelima**
# Add Blueprint System for Task Composition and Reusability

## Overview

This PR introduces a **Blueprint System** to NornFlow, enabling users to compose reusable task collections that can be referenced and expanded within workflows. Blueprints support variable resolution, conditional inclusion, nested composition, and circular dependency detection.

### 1. Blueprint Architecture

- **`BlueprintResolver`**: Handles variable context building and Jinja2 template resolution during workflow assembly
- **`BlueprintExpander`**: Orchestrates recursive blueprint expansion with circular dependency detection
- **Blueprint Files**: Simple YAML files containing only a tasks list that can be referenced by name or path

### 2. Variable Resolution Context

Blueprints support assembly-time variable precedence (lowest to highest):
1. Environment Variables (`NORNFLOW_VAR_*`)
2. Global Defaults (defaults.yaml)
3. Domain Defaults (`vars/{domain}/defaults.yaml`)
4. Workflow Variables (`workflow.vars`)
5. CLI Variables (`--vars`)

### 3. Conditional Blueprint Inclusion

Blueprints can be conditionally included using the `if` field:

```yaml
tasks:
  - blueprint: production_checks.yml
    if: "{{ environment == 'prod' }}"
  
  - blueprint: dev_setup.yml
    if: "{{ environment == 'dev' }}"
```

### 4. Nested Blueprint Composition

Blueprints can reference other blueprints, enabling modular task organization:

```yaml
# parent.yaml
tasks:
  - name: setup_task
  - blueprint: child_blueprint.yaml
  
# child_blueprint.yaml
tasks:
  - name: child_task_1
  - name: child_task_2
```

### 5. Circular Dependency Detection

The expander tracks blueprint expansion chains and raises `BlueprintCircularDependencyError` when circular references are detected, preventing infinite loops.

#### New Modules
- __init__.py - Blueprint package exports
- expander.py - Blueprint expansion logic
- resolver.py - Variable resolution and template rendering
- jinja2_utils.py - Centralized Jinja2 environment management

#### Modified Modules
- workflow.py - Integrated blueprint expansion in `WorkflowModel.create()`
- nornflow.py - Added blueprints catalog loading
- settings.py - Added `local_blueprints` configuration field
- show.py - Added `--blueprints` display option
- utils.py - Added `get_file_content_hash()` and renamed `is_workflow_file()` to `is_yaml_file()`
- mixins.py - Enhanced Jinja2 validation with syntax error detection

#### Settings
New configuration field in nornflow.yaml:
```yaml
local_blueprints:
  - "blueprints"  # Default directory
```

#### Show Command
```bash
# Display blueprints catalog
nornflow show --blueprints

# Display all catalogs (now includes blueprints)
nornflow show --catalogs
```

#### New Exceptions
- `BlueprintError` - Base exception for blueprint-related errors
- `BlueprintCircularDependencyError` - Specific error for circular dependencies

#### Jinja2 Environment Management
Centralized `Jinja2EnvironmentManager` class:
- Single source of truth for Jinja2 configuration
- Automatic registration of NornFlow custom filters
- Consistent error handling across the codebase

#### File Hashing
`get_file_content_hash()` generates stable content-based hashes for blueprints:
- Normalizes YAML formatting before hashing
- Used for circular dependency detection
- Returns 16-character hex strings

#### Blueprint Expander Tests (test_blueprint_expander.py)
- Simple blueprint expansion
- Conditional blueprint inclusion (`if` field)
- Nested blueprint composition
- Circular dependency detection
- Missing blueprint handling
- Invalid YAML structure detection

#### Blueprint Resolver Tests (test_blueprint_resolver.py)
- Variable context building with all sources
- Variable precedence validation
- Jinja2 template resolution
- Condition evaluation (boolean, string, Jinja2 expressions)
- Environment variable loading
- Domain defaults extraction

#### Workflow Model Tests (test_workflow_model.py)
- Blueprint expansion during workflow creation
- Mixed blueprints and direct tasks
- Variable resolution in blueprint references
- CLI variable integration
- Error handling for invalid blueprints

#### Utils Tests (test_utils.py)
- File content hashing (YAML normalization, consistency)
- YAML file validation
- All existing utility functions

#### Settings Tests (test_settings.py)
- `local_blueprints` default values
- Single and multiple blueprint directories
- Path resolution

### Test Statistics
- **10 new test files/classes**
- **60+ new test cases**
- **Comprehensive edge case coverage**

### Basic Blueprint

```yaml
# blueprints/device_checks.yaml
tasks:
  - name: check_version
    task: netmiko_send_command
    args:
      command_string: "show version"
  
  - name: check_interfaces
    task: netmiko_send_command
    args:
      command_string: "show interfaces status"
```

### Workflow Using Blueprint

```yaml
# workflows/health_check.yaml
workflow:
  name: "Device Health Check"
  tasks:
    - blueprint: device_checks.yaml
    - name: save_results
      task: write_file
      args:
        filename: "results.txt"
```

### Conditional Blueprint with Variables

```yaml
workflow:
  name: "Environment-Specific Deployment"
  vars:
    environment: "prod"
  tasks:
    - blueprint: production_validations.yaml
      if: "{{ environment == 'prod' }}"
    
    - blueprint: standard_deployment.yaml
    
    - blueprint: dev_cleanup.yaml
      if: "{{ environment == 'dev' }}"
```

### Dynamic Blueprint Resolution

```yaml
workflow:
  name: "Dynamic Blueprint Selection"
  vars:
    device_type: "router"
  tasks:
    - blueprint: "{{ device_type }}_config.yaml"  # Resolves to "router_config.yaml"
```

## Breaking Changes

**None** - This is a fully backward-compatible addition. Existing workflows without blueprints continue to work unchanged.

## Migration Guide

No migration required. To adopt blueprints:

1. Create a blueprints directory (or configure custom path in nornflow.yaml)
2. Create blueprint YAML files with tasks lists
3. Reference blueprints in workflows using `blueprint: <file_name_with_extension>` instead of defining tasks inline

## Documentation Updates

- Updated nornflow.yaml sample with `local_blueprints` field
- Enhanced CLI help text for `show` command
- Added blueprint descriptions to catalog display

### CODE REVIEW

File: nornflow/blueprints/expander.py [hunk 1]
```diff
@@ -0,0 +1,278 @@
...
+        blueprint_name = blueprint_ref.get("blueprint")
+        if not blueprint_name:
+            raise BlueprintError("Blueprint reference missing 'blueprint' field")
+
+        resolved_name = self.resolver.resolve_template(blueprint_name, context)
+        blueprint_path = self._resolve_blueprint_to_path(resolved_name, blueprints_catalog)
+        content_hash = get_file_content_hash(blueprint_path)
+
+        if content_hash in expansion_stack:
+            raise BlueprintCircularDependencyError(blueprint_path.name, name_stack)
```
[the remark this answers is not in the document]
Comment (theandrelima) [reply nornflow/blueprints/expander.py]: This is exactly the intended behavior. If two blueprints have different names, but same contents, then they are essentially the same blueprint, and we could throw ourselves in a loop here.

File: nornflow/models/workflow.py [hunk 2]
```diff
@@ -30,40 +32,65 @@ def create(cls, dict_args: dict[str, Any], *args: Any, **kwargs: Any) -> "Workfl
...
+
+        expanded_tasks = expander.expand_blueprints(
+            tasks=workflow_dict["tasks"],
+            blueprints_catalog=blueprints_catalog,
+            vars_dir=vars_dir,
+            workflow_path=workflow_path,
+            workflow_roots=workflow_roots,
+            inline_vars=workflow_dict.get("vars"),
+            cli_vars=cli_vars,
+        )
```
[the remark this answers is not in the document]
Comment (theandrelima) [reply nornflow/models/workflow.py]: this is ender consideration, as a BluePrint specific model.
however, this is not considered critical and won't be implemented at this point.
